File: offline-1/ScopeTable.cpp
#include <bits/stdc++.h>
#include "SymbolInfo.cpp"
#include "HashFunction.cpp"
using namespace std;

class ScopeTable
{
    SymbolInfo **hash_table;
    ScopeTable *parentScope;
    int num_buckets;
    int scope_id;
    HashFunction::HashFunc hashFunction; // Function pointer for the hash function

public:
    // Constructor that takes the hash function name as a string
    ScopeTable(int num_buckets, int scope_id, const string &hashFuncName = "SDBM")
    {
        cout << "\tScopeTable# " << scope_id << " created\n";
        this->num_buckets = num_buckets;
        this->scope_id = scope_id;
        this->parentScope = NULL;
        this->hashFunction = HashFunction::getHashFunction(hashFuncName); // Set hash function

        hash_table = new SymbolInfo *[num_buckets];
        for (int i = 0; i < num_buckets; i++)
        {
            hash_table[i] = NULL;
        }
    }

    ~ScopeTable()
    {
        for (int i = 0; i < num_buckets; i++)
        {
            if (hash_table[i] != NULL)
            {
                delete hash_table[i];
            }
        }
        delete[] hash_table;
        cout << "\tScopeTable# " << this->scope_id << " removed\n";
    }

    ScopeTable *getParentScope()
    {
        return this->parentScope;
    }

    void setParentScope(ScopeTable *s)
    {
        this->parentScope = s;
    }

    void setHashFunction(const string &hashFuncName)
    {
        this->hashFunction = HashFunction::getHashFunction(hashFuncName);
    }

    // Modify insertSymbol to use the hashFunction object
    bool insertSymbol(SymbolInfo s)
    {
        SymbolInfo *found = this->lookUp(s, false);
        if (found != NULL)
        {
            cout << "\t'" << s.getName() << "' already exists in the current ScopeTable\n";
            return false;
        }

        unsigned long long hash = hashFunction(s.getName(), this->num_buckets);
        int index = hash % num_buckets;
        int chain_position = 0;

        if (hash_table[index] == NULL)
        {
            hash_table[index] = new SymbolInfo(s);
        }
        else
        {
            chain_position += 1;
            SymbolInfo *temp = hash_table[index];
            while (temp->next != NULL)
            {
                temp = temp->next;
                chain_position += 1;
            }
            temp->next = new SymbolInfo(s);
        }
        cout << "\tInserted in ScopeTable# " << this->scope_id << " at position " << index + 1
             << ", " << chain_position + 1 << "\n";
        return true;
    }

    // Modify lookUp to use the hashFunction object
    SymbolInfo *lookUp(SymbolInfo s, bool showOutput = true)
    {
        string name = s.getName();
        unsigned long long hash = hashFunction(name, this->num_buckets) % num_buckets;
        int index = hash % num_buckets;
        int chain_position = 0;
        SymbolInfo *temp = hash_table[index];
        while (temp != NULL)
        {
            chain_position++;
            if (temp->getName() == name)
            {
                if (showOutput)
                {
                    cout << "\t'" << name << "' found in ScopeTable# " << this->scope_id
                         << " at position " << index + 1 << ", " << chain_position << "\n";
                }
                return temp;
            }
            temp = temp->next;
        }

        return NULL;
    }

    bool deleteSymbol(SymbolInfo s)
    {
        int index = hashFunction(s.getName(), this->num_buckets) % num_buckets;
        if (hash_table[index] == NULL)
        {
            cout << "\tNot found in the current ScopeTable\n";
            return false; // symbol not found
        }

        SymbolInfo *head = hash_table[index];
        int chain_position = 0;
        // if found at head, delete
        if (head->getName() == s.getName())
        {
            hash_table[index] = head->next; // delete the head node
            delete head;
            cout << "\tDeleted '" << s.getName() << "' from ScopeTable# " << this->scope_id
                 << " at position " << index + 1 << ", " << chain_position + 1 << "\n";

            return true; // symbol deleted successfully
        }
        // else traverse the chain
        else
        {
            while (head->next != NULL)
            {
                SymbolInfo *prev = head;
                head = head->next;
                chain_position++;
                if (head->getName() == s.getName())
                {
                    prev->next = head->next;
                    delete head;
                    cout << "\tDeleted '" << s.getName() << "' from ScopeTable# " << this->scope_id
                         << " at position " << index + 1 << ", " << chain_position + 1 << "\n";
                    return true; // symbol deleted successfully
                }
            }
            // if at the last elem
            if (head != NULL && head->getName() == s.getName())
            {
                delete head;
                cout << "\tDeleted '" << s.getName() << "' from ScopeTable# " << this->scope_id
                     << " at position " << index + 1 << ", " << chain_position + 1 << "\n";

                return true; // symbol deleted successfully
            }
        }
    }

    void print(int indentLevel = 0)
    {
        string indent(indentLevel, '\t');
        cout << indent << "ScopeTable# " << scope_id << "\n";
        for (int i = 0; i < num_buckets; i++)
        {
            cout << indent << i + 1 << "--> ";
            SymbolInfo *temp = hash_table[i];
            if (temp != NULL)
            {
                cout << temp->showSymbol() << " ";
                while (temp->next != NULL)
                {
                    temp = temp->next;
                    cout << temp->showSymbol() << " ";
                }
            }
            cout << "\n";
        }
    }
};

// TESTING CODE
// int main()
// {
//     //  Create a ScopeTable with 5 buckets and scope ID 1
//     ScopeTable scopeTable(5, 1);

//     // Insert symbols into the ScopeTable
//     SymbolInfo symbol1("x NUMBER");
//     SymbolInfo symbol2("y NUMBER");
//     SymbolInfo symbol3("z FUNCTION INT INT FLOAT");
//     SymbolInfo symbol4("a STRUCT INT id STRING name");

//     cout << "Inserting symbols:\n";
//     scopeTable.insertSymbol(symbol1);
//     scopeTable.insertSymbol(symbol2);
//     scopeTable.insertSymbol(symbol3);
//     scopeTable.insertSymbol(symbol4);

//     // Print the ScopeTable
//     cout << "\nPrinting ScopeTable:\n";
//     scopeTable.print();

//     // Look up symbols
//     cout << "\nLooking up symbols:\n";
//     SymbolInfo *foundSymbol = scopeTable.lookUp(symbol1);
//     if (foundSymbol)
//         cout << "Found: " << foundSymbol->showSymbol() << "\n";
//     else
//         cout << "Symbol not found\n";

//     foundSymbol = scopeTable.lookUp(SymbolInfo("nonexistent NUMBER"));
//     if (foundSymbol)
//         cout << "Found: " << foundSymbol->showSymbol() << "\n";
//     else
//         cout << "Symbol not found\n";

//     // Delete a symbol
//     cout << "\nDeleting symbols:\n";
//     scopeTable.deleteSymbol(symbol4);                          // Delete symbol 'y'
//     scopeTable.deleteSymbol(SymbolInfo("nonexistent NUMBER")); // Try to delete a non-existent symbol

//     // Print the ScopeTable after deletion
//     cout << "\nPrinting ScopeTable after deletion:\n";
//     scopeTable.print(1);
//     // cout << "NAFIS NAHIAN\n";
//     // cout << "\tNAFIS NAHIAN\n";
//     return 0;
// }

File: offline-4-ICG/icg_mine/SymbolTable/ScopeTable.java
package SymbolTable;

import java.io.*;

public class ScopeTable {
    private SymbolInfo[] hashTable;
    private ScopeTable parentScope;
    private int numBuckets;
    private String scopeId;
    private BufferedWriter out;

    // Constructor
    public ScopeTable(int numBuckets, String scopeId, BufferedWriter outputStream) {
        this.numBuckets = numBuckets;
        this.scopeId = scopeId;
        this.parentScope = null;
        this.out = outputStream;

        hashTable = new SymbolInfo[numBuckets];
        for (int i = 0; i < numBuckets; i++) {
            hashTable[i] = null;
        }
    }

    // Constructor with System.out as default output
    public ScopeTable(int numBuckets, String scopeId) {
        this(numBuckets, scopeId, new BufferedWriter(new OutputStreamWriter(System.out)));
    }

    // Getters and setters
    public ScopeTable getParentScope() {
        return this.parentScope;
    }

    public void setParentScope(ScopeTable s) {
        this.parentScope = s;
    }
    
    public String getScopeId() {
        return this.scopeId;
    }

    // Insert symbol into the scope table
    public boolean insertSymbol(SymbolInfo s) {
        SymbolInfo found = this.lookUp(s, false);
        if (found != null) {
            return false;
        }

        int hash = HashFunction.sdbmHash(s.getName());
        int index = hash % numBuckets;
        // int chainPosition = 0;

        if (hashTable[index] == null) {

            hashTable[index] = new SymbolInfo(s);
            try {
                String info = hashTable[index].showSymbol();
                System.out.println("Inserted symbol: " + info);
            } catch (Exception e) {
                // TODO: handle exception
            }

        } else {
            // chainPosition += 1;
            SymbolInfo temp = hashTable[index];
            while (temp.next != null) {
                temp = temp.next;
                // chainPosition += 1;
            }
            temp.next = new SymbolInfo(s);
        }
        return true;
    }

    // Look up symbol in the scope table
    public SymbolInfo lookUp(SymbolInfo s, boolean showOutput) {
        String name = s.getName();
        int hash = HashFunction.sdbmHash(s.getName());
        int index = hash % numBuckets;
        // int chainPosition = 0;
        SymbolInfo temp = hashTable[index];
        
        while (temp != null) {
            // chainPosition++;
            if (temp.getName().equals(name)) {
                return temp;
            }
            temp = temp.next;
        }
        return null;
    }

    // Look up symbol and return position information
    public SymbolInfo lookUp(SymbolInfo s, int[] foundIndex, int[] chainPos, boolean showOutput) {
        String name = s.getName();
        int hash = HashFunction.sdbmHash(s.getName());
        int index = hash % numBuckets;
        // int chainPosition = 0;
        SymbolInfo temp = hashTable[index];
        
        while (temp != null) {
            // chainPosition++;
            if (temp.getName().equals(name)) {
                foundIndex[0] = index;
                // chainPos[0] = chainPosition;
                return temp;
            }
            temp = temp.next;
        }
        foundIndex[0] = -1;
        chainPos[0] = -1;
        return null;
    }

    // Delete symbol from the scope table
    public boolean deleteSymbol(SymbolInfo s) {
        int hash = HashFunction.sdbmHash(s.getName());
        int index = hash % numBuckets;
        
        if (hashTable[index] == null) {
            return false; // symbol not found
        }

        SymbolInfo head = hashTable[index];
        // int chainPosition = 0;
        
        // if found at head, delete
        if (head.getName().equals(s.getName())) {
            hashTable[index] = head.next; // delete the head node
            return true; // symbol deleted successfully
        }
        // else traverse the chain
        else {
            while (head.next != null) {
                SymbolInfo prev = head;
                head = head.next;
                // chainPosition++;
                if (head.getName().equals(s.getName())) {
                    prev.next = head.next;
                    return true; // symbol deleted successfully
                }
            }
            // if at the last elem
            if (head != null && head.getName().equals(s.getName())) {
                return true; // symbol deleted successfully
            }
        }
        return false; // symbol not found
    }

    // Print the scope table
    public void print(int indentLevel) {
        StringBuilder indent = new StringBuilder();
        for (int i = 0; i < indentLevel; i++) {
            indent.append('\t');
        }
        
        try {
            out.write(indent + "ScopeTable # " + scopeId + "\n");
            for (int i = 0; i < numBuckets; i++) {
                SymbolInfo temp = hashTable[i];
                if (temp != null) {
                    out.write(indent.toString());
                    out.write(i + " --> ");
                    while (temp != null) {
                        // out.write("< " + temp.getName() + " : " + temp.getType() + 
                        //           ", Offset: " + temp.offset + " > ");
                        out.write(temp.showSymbol());
                        temp = temp.next;
                    }
                    out.write("\n");
                }
            }
            out.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void print() {
        print(0);
    }
}
